docs/experiments/extract-mode.md

```diff
@@ -0,0 +1,103 @@
+---
+id: extract-mode
+append_help_link: true
+description: "This article explains the extract mode, which allows for easier handling of files containing more than one language."
+---
+
+# Extract mode
+
+## Introduction
+
+Extract mode enables you to run existing rules on subsections of files where the rule language is different than the language of the file. For example, running a Javascript rule on code contained inside of script tags in an HTML document.
+
+## Example of extract mode
+
+Without extract mode, writing rules to validate template, Markdown or configuration files which contain code in another language can be burdensome and require significant rule duplication.
+
+Let's take the following Bash rule as an example (a simplified version of [our `curl-eval` rule](https://github.com/returntocorp/semgrep-rules/blob/release/bash/curl/security/curl-eval.yaml)):
+
+```yaml
+rules:
+  - id: curl-eval
+    severity: WARNING
+    languages:
+      - bash
+    message: Evaluating data from a `curl` command is unsafe.
+    mode: taint
+    pattern-sources:
+      - pattern: |
+          $(curl ...)
+      - pattern: |
+          `curl ...`
+    pattern-sinks:
+      - pattern: eval ...
+```
+
+Usually, this rule would only be run against Bash files. However, our project might contain Dockerfiles or Python scripts which invoke Bash commands&mdash;without an extract mode rule Semgrep won't know about any contained Bash, and therefore, will **not** run any Bash rules against those contained commands.
```

Comment:
Blocking:
- Avoid would, should, may - see https://developers.google.com/style/word-list?hl=en#would
  <img width="896" alt="Screenshot 2022-07-12 at 23 03 21" src="https://user-images.githubusercontent.com/26669307/178594860-f17bd1f8-a105-4140-8610-812968544aa2.png">
- Avoid software anthropomorphism - Semgrep cannot know anything by itself. See [Avoiding software anthropomorphism](https://www.notion.so/r2cdev/r2c-technical-communication-style-guide-4d1104e78927460aa8c7624aedd6561f#1d51679a2882492abaf13f4ed2b822ea) in r2c technical communication style guide.
- Do not use "we", "our" and similar words unless really necessary. See: [Use of we](https://www.notion.so/r2cdev/r2c-technical-communication-style-guide-4d1104e78927460aa8c7624aedd6561f#39871bd6aca04e37bac6dc2e70dd2baf) in r2c technical communication style guide.
- Avoid [will](https://developers.google.com/style/word-list?hl=en#will) when possible. Documentation needs to be written mostly in the present simple tense: See will in the Google developer style guide.Avoid [will](https://developers.google.com/style/word-list?hl=en#will) when possible. Documentation needs to be written mostly in the present simple tense: See will in the Google developer style guide. 

```suggestion
Usually, Semgrep uses this rule only against Bash files. However, a project might contain Dockerfiles or Python scripts that invoke Bash commands&mdash;without an extract mode rule, Semgrep does **not** run any Bash rules against commands contained in files of different languages.
```

Reply:
I probably should write rules for this...